src/v/kafka/protocol/schemata/generator.py

```diff
@@ -1047,18 +1086,115 @@ class response;
 {%- endif %}
 {%- endmacro %}
 
-{% macro struct_serde(struct, field_serde, flex, obj = "") %}
+{% macro blind_decode() %}
+reader.consume_tags();
+{%- endmacro %}
+
+{% macro encode_zero() %}
+writer.write_tags();
+{%- endmacro %}
+
+{% macro tag_decoder_impl(tag_definitions, obj = "") %}
+/// Tags decoding section
+auto num_tags = reader.read_unsigned_varint();
+while(num_tags-- > 0) {
+    auto tag = reader.read_unsigned_varint();
+    auto sz = reader.read_unsigned_varint(); // size
+    switch(tag){
+{%- for tdef in tag_definitions %}
+    case {{ tdef.tag() }}:
+{{- field_decoder(tdef, (field_decoder, tag_decoder), obj) | indent | indent }}
+        break;
+{%- endfor %}
+    default:
+        /// Read unknown tag
+        reader.consume_unknown_tag(sz);
+    }
+}
+{%- endmacro %}
+
+{% macro tag_decoder(tag_definitions, obj = "") %}
+{%- if tag_definitions|length == 0 %}
+{{- blind_decode() }}
+{%- else %}
+{
+{{- tag_decoder_impl(tag_definitions, obj) | indent }}
+}
+{%- endif %}
+{%- endmacro %}
+
+{% macro conditional_tag_encode(tdef, vec) %}
+{%- if tdef.is_array %}
+{%- if tdef.nullable() %}
+{%- call tag_version_guard(tdef) %}
+if ({{ tdef.name }}) {
+    {{ vec }}.push_back({{ tdef.tag() }});
+}
+{%- endcall %}
+{%- else %}
+{%- call tag_version_guard(tdef) %}
+if (!{{ tdef.name }}.empty()) {
+    {{ vec }}.push_back({{ tdef.tag() }});
+}
+{%- endcall %}
+{%- endif %}
+{%- elif tdef.default_value() != "" %}
+{%- call tag_version_guard(tdef) %}
+if ({{ tdef.name }} != {{ tdef.default_value() }}) {
+    {{ vec }}.push_back({{ tdef.tag() }});
+}
+{%- endcall %}
+{%- endif %}
+{%- endmacro %}
+
+{% macro tag_encoder_impl(tag_definitions, obj = "") %}
+/// Tags encoding section
+std::vector<uint32_t> to_encode;
+{%- for tdef in tag_definitions -%}
+{{- conditional_tag_encode(tdef, "to_encode") }}
+{%- endfor %}
+writer.write_unsigned_varint(to_encode.size());
+for(size_t tag : to_encode) {
```

Comment:
Is this a valid use of `write_flex_array`?

Reply:
not really because `write_flex_array` expects a list of homogeneous type T, while this code serializes a different type depending on the tag id